### PR TITLE
[SPARK-14244][SQL] Don't use SizeBasedWindowFunction.n created on executor side when evaluating window functions

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

`SizeBasedWindowFunction.n` is a global singleton attribute created for evaluating size based aggregate window functions like `CUME_DIST`. However, this attribute gets different expression IDs when created on both driver side and executor side. This PR adds `withPartitionSize` method to `SizeBasedWindowFunction` so that we can easily rewrite `SizeBasedWindowFunction.n` on executor side.
## How was this patch tested?

A test case is added in `HiveSparkSubmitSuite`, which supports launching multi-process clusters.

### CODE REVIEW

**Comment by liancheng** [comment]
Just came to a simpler and less error prone solution after opening this PR: we may reserve an expression ID range, say 0 ~ 1024, for predefined global attribute references like `SizeBasedWindowFunction.n`, and then create these attribute references with fixed expression IDs.

Will try the above approach tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 1]
```diff
@@ -937,6 +939,18 @@ private[execution] object AggregateProcessor {
       imperatives.toArray,
       trackPartitionSize)
   }
+
+  // Tries to find and return `SizeBasedWindowFunction.n` generated on driver side from given
+  // expressions.  Returns executor side global `SizeBasedWindowFunction.n` if none can be found.
+  def windowPartitionSizeAttribute(expressions: Seq[Expression]): AttributeReference = {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: `SizeBasedWindowFunction` has an `n` function which returns this attribute. Why not use this?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Ah, that's right! Not quite familiar with windowing code path. Thanks for the pointer!

**Comment by liancheng** [comment]
Reconsidered the 2nd approach mentioned above and found that it's not doable, since we also included JVM ID (a UUID) in `ExprId`. That said, binding `AttributeReference`s across different JVM processes is now allowed.

Took @hvanhovell's advice and added `withPartitionSize` to `SizeBasedWindowFunction` so that we can easily rewrite the partition size attribute reference on executor side.

A test case is added in `HiveSparkSubmitSuite`, which supports launching multi-process clusters.

**Comment by liancheng** [comment]
@yhuai This is ready for review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 2]
```diff
@@ -937,6 +944,14 @@ private[execution] object AggregateProcessor {
       imperatives.toArray,
       trackPartitionSize)
   }
+
+  // Rewrites the window partition size attribute in all `SizeBasedWindowFunction`s using executor
+  // side `SizeBasedWindowFunction.n`.
+  def rewriteWindowPartitionSize(expression: Expression): Expression = {
+    expression.transformDown {
+      case f: SizeBasedWindowFunction => f.withPartitionSize(SizeBasedWindowFunction.n)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Minor: we could also create a BoundReference directly. It is always at position 0.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 3]
```diff
@@ -885,9 +886,15 @@ private[execution] object AggregateProcessor {
...
+    // SPARK-14244: `SizeBasedWindowFunction`s created on driver side and serialized to executor
+    // side contains global window partition size attribute reference (`SizeBasedWindowFunction.n`),
+    // and must be rewritten using the one instantiated on executor side. Otherwise, attribute
+    // binding fails because of different expression ID.
+    val rewrittenFunctions = functions.map(rewriteWindowPartitionSize)
+
     // Check if there are any SizeBasedWindowFunctions. If there are, we add the partition size to
     // the aggregation buffer. Note that the ordinal of the partition size value will always be 0.
-    val trackPartitionSize = functions.exists(_.isInstanceOf[SizeBasedWindowFunction])
+    val trackPartitionSize = rewrittenFunctions.exists(_.isInstanceOf[SizeBasedWindowFunction])
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Minor: We could alo integrate this with rewriting if you don't mind side effects.

**Comment by hvanhovell** [comment]
Two minor suggestions. LGTM pending tests.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 3 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Actually I have to resort to side effects as the extra `Expression` constructor argument fails function lookup and the function registry always falls back to Hive `GenericUDAFCumeDist`. That's the reason of the most recent test failures.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 2 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: I'd prefer not to do so since it doesn't bring much performance gain, yet depends on a not so obvious implicit assumption (partition size attribute ordinal is always 0) and makes the code harder to grasp.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 4]
```diff
@@ -885,6 +886,12 @@ private[execution] object AggregateProcessor {
     val evaluateExpressions = mutable.Buffer.fill[Expression](ordinal)(NoOp)
     val imperatives = mutable.Buffer.empty[ImperativeAggregate]
 
+    // SPARK-14244: `SizeBasedWindowFunction`s created on driver side and serialized to executor
+    // side contains global window partition size attribute reference (`SizeBasedWindowFunction.n`),
+    // and must be rewritten using the one instantiated on executor side. Otherwise, attribute
+    // binding fails because of different expression ID.
+    functions.foreach(rewriteWindowPartitionSize)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: I think it's better to use the `n` from driver, instead of using the one created in executor. That could be found in any SizeBasedWindowFunction

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: After rethinking about it, reusing driver side attribute indeed makes more sense, which is also what the first commit of this PR does.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 5]
```diff
@@ -451,7 +451,7 @@ abstract class RowNumberLike extends AggregateWindowFunction {
  * A [[SizeBasedWindowFunction]] needs the size of the current window for its calculation.
  */
 trait SizeBasedWindowFunction extends AggregateWindowFunction {
-  protected def n: AttributeReference = SizeBasedWindowFunction.n
+  val n: AttributeReference = SizeBasedWindowFunction.n
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: It's made a `val` so that the attribute created on driver side is serialized to executor side.

**Comment by liancheng** [comment]
@hvanhovell @davies Thanks for the review!

**Comment by davies** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 5 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: How about we add this kind of comments to the code in future? For this one, we can add it when we check it in.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 6]
```diff
@@ -937,6 +945,12 @@ private[execution] object AggregateProcessor {
       imperatives.toArray,
       trackPartitionSize)
   }
+
+  // SPARK-14244: When binding
+  def windowPartitionSizeAttribute(expressions: Array[Expression]): AttributeReference = {
+    val aggs = expressions.flatMap(_.collectFirst { case f: SizeBasedWindowFunction => f })
+    aggs.headOption.map(_.n).getOrElse(SizeBasedWindowFunction.n)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: When aggs will be empty?

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: When there are no SizeBasedWindowFunctions...

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: If we do not have any `SizeBasedWindowFunction`, should this method be called? When will we need to use `SizeBasedWindowFunction.n` created at the executor side? I guess I am confused why we need https://github.com/apache/spark/pull/12040/files#diff-5dade93a3cce85de2ceeab02bd9d4ff6R931.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: If there's no `SizeBasedWindowFunction`, actually we don't need the window partition size attribute, and invoking this function can be viewed as a no-op.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 5 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: Just added the comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 6 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: If this method will be called from a single place, should we inline it?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Hm, makes sense. And, I just realized that I didn't complete the comment...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 7]
```diff
@@ -920,7 +931,7 @@ private[execution] object AggregateProcessor {
     // Create the projections.
     val initialProjection = newMutableProjection(
       initialValues,
-      Seq(SizeBasedWindowFunction.n))()
+      Seq(partitionSize))()
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: @hvanhovell  Why we only set a single column (partitionSize) as the input schema? I guess I am missing something...

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: @yhuai One of the `SizeBasedWindowFunction`s (NTILE) uses the `partitionSize` during initialization. So we need have it available as an input attribute during initialization.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 8]
```diff
@@ -885,11 +886,21 @@ private[execution] object AggregateProcessor {
     val evaluateExpressions = mutable.Buffer.fill[Expression](ordinal)(NoOp)
     val imperatives = mutable.Buffer.empty[ImperativeAggregate]
 
+    // SPARK-14244: `SizeBasedWindowFunction`s are firstly created on driver side and then
+    // serialized to executor side. These functions all reference a global singleton window
+    // partition size attribute reference, i.e., `SizeBasedWindowFunction.n`. Here we must collect
+    // the singleton instance created on driver side instead of using executor side
+    // `SizeBasedWindowFunction.n` to avoid binding failure caused by mismatching expression ID.
+    val partitionSize = {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: @liancheng how about we return the option here? Then we don't need to do another iteration round to determine the value of track partitions, and use `isDefined`/`foreach` and `get`/`toSeq` further down the line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 7 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: The actual buffer will have more values, right? We always put partitionSize at the first value? Or, if we do not have a sized based function, we will later set the first value again?

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Yeah the initial buffer will have more values. A `partitionSize` attribute will only be created/initialized when a `SizeBasedWindowFunction` is used. The `partitionSize` will be put at the first position, and others will be placed after that.

The initialization logic can be found here: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala#L965-L967

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: OK I see. I think I am confused about the reason that the inputSchema of this mutable projection only contains a single column `partitionSize`. Does initialValues already have all other needed BoundReferences?

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: An initial value is allmost allways a constant. So in that case you don't need an input variable. `partitionSize` is a bit special because we know the value before the aggregates get initialized but not early enough to efficiently employ somekind of folding.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: OK. Thanks! How about we add comment to explain this. This logic is not very easy to understand. Because in most of cases (when `trackPartitionSize=false`), we do not need to bind when we set initial values (because most of them will be constant values), we just put partitionSize at here.

So, the logical is actually the following, right? 
```
val initialProjection =
  if (trackPartitionSize) {
    newMutableProjection(initialValues, Seq(partitionSize))()
  } else {
    newMutableProjection(initialValues, Seq())()
  }
```

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Yeah it is.

Further documentation never hurts.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 8 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Yeah, saw discussion between you and Yin above. This improves readability. TBH, I was also quite confused while updating this part of code. Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 9]
```diff
@@ -107,7 +107,9 @@ private[hive] class HiveFunctionRegistry(
           // If there is any other error, we throw an AnalysisException.
           val errorMessage = s"No handler for Hive udf ${functionInfo.getFunctionClass} " +
             s"because: ${throwable.getMessage}."
-          throw new AnalysisException(errorMessage)
+          val analysisException = new AnalysisException(errorMessage)
+          analysisException.setStackTrace(throwable.getStackTrace)
+          throw analysisException
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: This change is not directly related to this PR, but it preserving the original exception stack trace helps debugging Hive function lookup here.

**Comment by yhuai** [comment]
LGTM!
